```
/Users/Jeremiah/Documents/ParkyParkPark/app/js/menus/VariantMenu.js
  3:10  error  "key" is defined but never used     no-unused-vars
  5:13  error  "levels" is defined but never used  no-unused-vars
```

Not edited here, so it must be on master. I'll just fix this in another commit.

--- a/app/js/ParkingSpace.js
+++ b/app/js/ParkingSpace.js
@@ -25,6 +25,8 @@ export class ParkingSpace {
       sensor:true
     });
 
+    this.body._entity = this
+
 
     this.texture = opts.texture;
     this.graphics = new PIXI.Graphics();
@@ -35,8 +37,9 @@ export class ParkingSpace {
     this.sprite.y = this.sprite.y + this.sprite.height/2
     this.sprite.rotation =3.14159
     this.boxShape = new p2.Box({ width: this.sprite.width, height: this.sprite.height });
-    // this.boxShape.collisionGroup = config.CAR;
-    // this.boxShape.collisionMask = config.PLAYER | config.CAR | config.TRUCKBACK | config.WALL | config.BOMB | config.EXPLOSION
+    this.boxShape.collisionGroup = config.SCORE;
+    this.boxShape.collisionMask = config.CAR | config.TRUCKBACK
+    this.boxShape.collsionResponse = false
     this.body.addShape(this.boxShape);
   }
 

--- a/app/js/config.js
+++ b/app/js/config.js
@@ -5,6 +5,7 @@ export default {
   zoom: 0.05*W,
   W,
   H,
+  SCORE: Math.pow(2,0),
   PLAYER: Math.pow(2,1),
   CAR: Math.pow(2,2),
   WALL: Math.pow(2,3),

--- a/app/js/entities/Car.js
+++ b/app/js/entities/Car.js
@@ -19,7 +19,7 @@ const defaults = {
   collisionGroup:config.PLAYER,
   stage:config.stage,
   texture:resources.RedCar.texture,
-  collisionMask:config.PLAYER | config.CAR | config.TRUCKBACK | config.WALL | config.BOMB | config.EXPLOSION,
+  collisionMask:config.PLAYER | config.CAR | config.TRUCKBACK | config.WALL | config.BOMB | config.EXPLOSION | config.SCORE,
   wheelTexture:resources.wheel.texture,
   wheelPositions: [
     { x:-0.22, y:0.24 },
@@ -36,6 +36,8 @@ export default class Car extends Entity {
     defaults.wheelTexture = resources.wheel.texture;
     opts = _.defaults(opts, defaults);
     super(opts)
+
+    this.score = 1
 
     this.body.onCollision = (body, otherShape, playerHit) => {
       this.setSideFriction(3,3);

--- a/app/js/entities/Entity.js
+++ b/app/js/entities/Entity.js
@@ -23,6 +23,8 @@ export default class Entity {
     });
     config.world.addBody(this.body);
 
+    this.body._entity = this
+
     this.graphics = new PIXI.Graphics();
     opts.container.addChild(this.graphics);
 

--- a/app/js/entities/Truck.js
+++ b/app/js/entities/Truck.js
@@ -24,6 +24,7 @@ export default class Truck extends Car {
     opts = _.defaults(opts, defaults);
     super(opts)
     this.exploded = false;
+    this.score = 3
 
     this.body.removeShape(this.boxShape);
     this.boxShape = new p2.Box({ width: opts.w-0.1, height: opts.h-0.675 });

--- a/app/js/gamemodes/valet.js
+++ b/app/js/gamemodes/valet.js
@@ -3,14 +3,16 @@ import config from '../config';
 import gamestate from '../gamestate';
 import * as Cars from '../entities/Vehicles.js';
 import MainMenu from '../menus/MainMenu.js'
-import menuLoop from '../menuLoop.js'
 
 import _ from 'lodash';
 
 export default class Valet extends GameMode {
-  constructor() {
+  constructor(cars) {
     super()
-    this.carsLeft = new PIXI.Text('Cars Left: 24',{ font : '24px Arial', fill : 0xFFFFFF, align : 'center' })
+
+    gamestate.carsLeft = cars
+
+    this.carsLeft = new PIXI.Text('Cars Left: ' + cars,{ font : '24px Arial', fill : 0xFFFFFF, align : 'center' })
     this.carsLeft.x = 0
     this.carsLeft.y = 0
     config.stage.addChildAt(this.carsLeft, config.stage.children.length)
@@ -39,12 +41,78 @@ export default class Valet extends GameMode {
         this.carsLeft.text = 'Cars Left: ' + gamestate.carsLeft
       }
 
-      if(gamestate.carsLeft === 0) {
+      if (gamestate.carsLeft === 0) {
         gamestate.playing = false;
+        // Game ends when there are no cars left.
+        // Reset the score so we can calculate it.
         gamestate.score = 0;
-        for (let i = 0; i < gamestate.level.parkingSpaces.length; i++) {
-          gamestate.score += gamestate.level.parkingSpaces[i].getScore();
+
+        // render the updated carsLeft already
+        config.renderer.render(config.stage);
+
+        // do some setup and add the parkingSpaces' bodies to the world
+        const psLookup = []
+        const parkingSpaces = gamestate.level.parkingSpaces
+        for (let i = 0; i < parkingSpaces.length; i++) {
+          const body = parkingSpaces[i].body
+          body._scores = []
+          config.world.addBody(body);
+          psLookup.push(body)
         }
+
+        // register an on beginContact so we have have P2 tell us which spaces have cars in them
+        config.world.on('beginContact', (payload) => {
+          // because eslint no-redecale can't handle exclusive if-statement hoisting
+          var psBody
+          var other
+          if (_.contains(psLookup, payload.bodyA)) {
+            psBody = payload.bodyA
+            other = payload.bodyB
+          } else if (_.contains(psLookup, payload.bodyB)) {
+            psBody = payload.bodyB
+            other = payload.bodyA
+          } else {
+            console.warn('Unexpected contact')
+            console.warn('Entities', payload.bodyA._entity, payload.bodyB._entity)
+            return
+          }
+
+          // some wild math suggested by schteppe (p2.js Author)
+          // This doesn't strictly check for overlap, but rather which body has
+          // the deepest penetration. i.e. Has the point on a shape the farthest
+          // from the farthest point on the other shape.
+          const contactEq = payload.contactEquations[0]
+          const penetrationVec = contactEq.penetrationVec
+          p2.vec2.add(penetrationVec,contactEq.contactPointB,contactEq.bodyB.position)
+          p2.vec2.sub(penetrationVec,penetrationVec,contactEq.bodyA.position)
+          p2.vec2.sub(penetrationVec,penetrationVec,contactEq.contactPointA);
+
+          const _score = {
+            depth: p2.vec2.dot(contactEq.penetrationVec, contactEq.normalA),
+            body: other
+          }
+          psBody._scores.push(_score)
+        })
+
+        // Step the world so that P@ can update and create contacts for the parkingSpaces
+        config.world.step(1/60)
+
+        // Actually calculate the score based on which vehicle has the most "overlap" for each space.
+        _.each(psLookup, (parkingSpace) => {
+          var depth = -0.2 // minimum overlap to get score
+          var body = null
+          _.each(parkingSpace._scores, (_score) => {
+            if (_score.depth >= depth) return
+
+            depth = _score.depth
+            body = _score.body
+          })
+
+          if (!body) return
+          gamestate.score += body._entity.score // get score from the entity type
+        })
+
+        // reset everything
         config.stage.removeChild(this.carsLeft)
         config.world.clear();
         config.world.gravity = [0,0];
@@ -56,7 +124,7 @@ export default class Valet extends GameMode {
         gamestate.menus.push(menu)
         config.stage.addChild(menu)
         console.log(gamestate.score)
-        requestAnimationFrame(menuLoop);
+        requestAnimationFrame(_loop);
       }
 
       for (let i = gamestate.cars.length - 1; i >= 0; i--) {

--- a/app/js/levels.js
+++ b/app/js/levels.js
@@ -18,11 +18,6 @@ class Level {
     this.texture = texture;
     this.spawnPoints = [];
     this.parkingSpaces = []
-
-    for (var i = 0; i < 24; i++) {
-      this.parkingSpaces.push(new ParkingSpace({ x:Random.integer(1,600)(mt) + 250, y:Random.integer(-400,-1)(mt) -40 , angle: Math.floor(Math.random() * (Math.PI*2)) }))
-
-    }
   }
 
   load() {
@@ -51,6 +46,10 @@ class Level {
     this.spawnPoints.push({ x, y, velX, velY })
   }
 
+  addParkingSpace(x, y, angle) {
+    this.parkingSpaces.push(new ParkingSpace({ x, y, angle }))
+  }
+
   debug(toggle) {
     for (let i = 0; i < this.walls.length; i++) {
       this.walls[i].debug(toggle)
@@ -70,6 +69,22 @@ export class Test extends Level {
     this.addWall(101,-300,203,333,0);
     this.addSpawn(config.scaleFactorX*-20, config.scaleFactorY*-95);
     this.addSpawn(config.scaleFactorX*-20, config.scaleFactorY*-500)
+
+    const up = 0
+    const down = 180 * (Math.PI / 180)
+
+    this.sprite = new PIXI.Sprite(resources.ParkingSpace.texture);
+
+    var psOffsetY = this.sprite.height / 2
+    var psOffsetX = this.sprite.width / 2
+
+    for (let i = 0; i < 8; i++) {
+      this.addParkingSpace((psOffsetX * i + 360), -150, down)
+      this.addParkingSpace((psOffsetX * i + 360), -150 - psOffsetY, up)
+
+      this.addParkingSpace((psOffsetX * i + 360), -370, up)
+      this.addParkingSpace((psOffsetX * i + 360), -370 + psOffsetY, down)
+    }
   }
 }
 

--- a/app/js/menus/MainMenu.js
+++ b/app/js/menus/MainMenu.js
@@ -18,8 +18,7 @@ export default class MainMenu extends Menu {
     this.addOption('Play', (menus) => {
       menus.splice(menus.indexOf(this))
       config.stage.removeChild(this)
-      gamestate.mode = new Valet();
-      gamestate.carsLeft = 24;
+      gamestate.mode = new Valet(4);
       return { _playing: true, _level: new variant.level() };
     });
 

